### DISCUSSION TITLE
Step changing selected step by code

### DISCUSSION THREAD

**Discussion by svenbaus** [Megabit/Blazorise#4104] [Q&A]
Hello,
I'm currently facing a bit strange behaviour. Setting the selected step by code works, but the GUI reacts not as expected. First my setup:

```
<Steps SelectedStep="@selectedStep" SelectedStepChanged="OnSelectedStepChanged">
	<Items>
		@foreach(var processStep in processFlow.ProcessSteps)
		{
			<Step Name="@processStep.Id.ToString()" Completed="GetStepCompleted(processStep)" Color="GetStepColor(processStep)">
				<Marker>
					@{
						var iconName = GetIconName(processStep);
						@if (iconName != null)
						{
							<Icon Name="@iconName" />
						}
					}
				</Marker>
				<Caption>
					@String.Format("{0} {1}", processStep.Position, _iProcessStepClassLocalizer[processStep.ProcessStepClass?.FullName])
				</Caption>
			</Step>
		}
	</Items>
	<Content>
		@foreach(var processStep in processFlow.ProcessSteps)
		{
			<StepPanel Name="@processStep.Id.ToString()">
				[...]
			</StepPanel>
		}
	</Content>
</Steps>
```

Updating the selected step by code:
```
selectedStep = processFlow.NextProcessStepToExecute.Id.ToString();
InvokeAsync(() => StateHasChanged());
```

The GUI looks after a change like this:
![grafik](https://user-images.githubusercontent.com/86463105/190104705-a353d61b-6fba-4c92-9382-ef8946c6aab4.png)

How can I make shure, that the step 2 is selected as active step, but doesn't change the positions?

**Comment by stsrki** [comment]
I'm not exactly sure what you mean. Selected steps and completed steps are different things.

You can optimize it by adding a key to the step in a foreach loop, eg `<Step @key="@processStep">`

**Reply by svenbaus** [reply]
Well basically the steps should be ordered like "1", "2", "3". Before an update of the selected step is done, this is the case. But if I update the selected step (from backend, running a process engine), the order becomes for example "2","1","3". Adding a `@key` didn't solve that problem. If I do a reload of the page, the order is correct again until I do the next update.

**Reply by stsrki** [reply]
Have you tried ordering it yourself?

eg,

`processFlow.ProcessSteps.Order(x=>x.something)`

**Reply by svenbaus** [reply]
Good idea, I debugged the system and found out, that the ordering of processSteps is wrong (the property itself is ordered the way it is displayed). So there is no bug on your side, but on my ;). Thanks anyway :).
